include/swift/AST/UndefDiagnosticMacros.h

```diff
@@ -0,0 +1,29 @@
+//===--- UndefDiagnosticMacros.h - Shared Diagnostic Macros -----*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file un-defines macros shared across diagnostic definition files.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef DIAG_NO_UNDEF
+
+#if defined(DIAG)
+#undef DIAG
+#endif
+
+#undef REMARK
+#undef NOTE
+#undef WARNING
+#undef ERROR
+#undef FIXIT
+
+#endif
```

Comment:
It may be a good idea to keep this in the same header with a define to indicate that you want them undefined.  That way they can be kept in sync.  That said, new levels are unlikely to be introduced, so perhaps the complexity isnt worth it.  Something to consider still.

Reply:
That's a good point. I'm still considering whether or not to extend these as part of some related diagnostics work. I think it's best to leave them as separate headers for now to keep things simple, but I'll reevaluate if it gets any more complicated.